Put for Rest API added

HenriBom opened the pull request.

HenriBom committed:
Add main file for project

HenriBom committed:
Put for Rest API added

cocap10 reviewed (changes requested):
You must add .idea to your gitignore in order to ignore all files under .idea dir

cocap10 reviewed (changes requested):
You must ignore all target dir too

HenriBom committed:
gitignore added

HenriBom committed:
.gitignore update

Commit:
Update .gitignore

Commit:
Merge pull request #2 from Zenika/update-gitignore

Update .gitignore

HenriBom committed:
merge

HenriBom committed:
bug in creation of foreign key i data base

HenriBom committed:
data base refactored. Set reciep by cotegory done. TODO : Get recipe from db

HenriBom committed:
Test: RecipeController

Add test for controller. ToDo : update and post

cocap10 reviewed (commented):
il y a encore des .idea et meme un  out/artifacts/zba_data_jar/zba-data.jar de comité.

cocap10 commented on src/test/java/org/zenika/zba/zbadata/RecipeDaoTest.java:
```diff
@@ -0,0 +1,55 @@
+        // when
+        Recipe found = recipeDao.findByName(toFind.getName());
+        // then
+        assertThat(found.getName())
```
pour info tu aurai pu utiliser:
assertThat(found).isEqualToComparingFieldByField(toFind)

cocap10 commented on src/test/java/org/zenika/zba/zbadata/RecipeControllerTest.java:
```diff
@@ -0,0 +1,216 @@
+    }
+}
+
+/**
```
avec git tu ne devrais pas avoir besoin de commenter du code

HenriBom replied:
C'est du code dupliqué. J'ai essayé de faire mes mock en utilisant WebMvcTest mais sans succès du coup  je l'ai gardé en commentaire pour avoir votre avis là dessus

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/RecipeController.java:
```diff
@@ -2,76 +2,72 @@
         List<Recipe> recipe = recipeDao.findAll();
-        if(recipe == null) throw new RecipeNotFindException("No recipe find");
-        return recipe;
+        if (recipe == null) throw new RecipeNotFindException("No recipe found");
```
`recipeDao.findAll()` ne renvoie jamais null mais liste vide quand il n'y a pas de Recipe

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/RecipeController.java:
```diff
@@ -2,76 +2,72 @@
-        Recipe recipe = recipeDao.findById(value);
-        if(recipe == null) throw new RecipeNotFindException("No recipe find for this id");
-        return recipe;
+    @GetMapping(value = "/Steps{id}")
```
c'est pas plutot `/steps/{id}` ?

cocap10 replied:
ou meme `recipe/{id}/seteps` ? car id c'est le recipeId

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/RecipeController.java:
```diff
@@ -2,76 +2,72 @@
+    @Autowired
+    private StepDao stepDao;
 
     @GetMapping(value = "/Recipe")
```
pour une API je pense que tu devrait laisser en minuscule donc `/recipe`

cocap10 replied:
tu peux mettre l'annotation au niveau classe, comme ici : https://stackoverflow.com/questions/22702568/spring-mvc-requestmapping-both-class-and-method pour éviter de répéter /recipe

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/RecipeController.java:
```diff
@@ -2,76 +2,72 @@
+        for (RecipeStep recipeStep: recipeSteps) {
+            steps.add(recipeStep.getStep());
+        }
+        if (steps == null) throw new StepsNotFindException("No step found");
```
steps ne sera jamais null car tu le set a la ligne 44.
Ce que tu veux faire c'est steps.isEmpty()

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/RecipeController.java:
```diff
@@ -2,76 +2,72 @@
 
-    @GetMapping(value = "/Recipe/name{value}")
-    public Recipe findRecipeName(@PathVariable String value) throws RecipeNotFindException {
+    @DeleteMapping(value = "/Recipe{value}")
```
/recipe/{value}

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/RecipeController.java:
```diff
@@ -2,76 +2,72 @@
-    @GetMapping(value = "/Recipe/name{value}")
-    public Recipe findRecipeName(@PathVariable String value) throws RecipeNotFindException {
+    @DeleteMapping(value = "/Recipe{value}")
+    public ResponseEntity<String> deleteRecipe(@PathVariable String value) throws RecipeNotFindException {
```
tu peux peut-etre faire un namedQuery avec `deleteByName`

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/RecipeController.java:
```diff
@@ -2,76 +2,72 @@
-        List<Recipe> recipe = recipeDao.findByCreator(value);
-        if(recipe == null) throw new RecipeNotFindException("No recipe find");
-        return recipe;
+        if (recipe == null) throw new RecipeNotFindException("No recipe found for this name");
```
dans l'erreur ce serai bien de préciser le nom de la recette qui n'a pas été trouvée

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/RecipeController.java:
```diff
@@ -2,76 +2,72 @@
+    @PostMapping(value = "/Recipe")
+    public Object addRecipe(@RequestBody Object object) {
+        Save save = new Save();
+        System.out.println("post");
```
Pour la suite ca serai bien d'utiliser un loggeur plutot que de faire des system.out

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/recipe/Save.java:
```diff
@@ -0,0 +1,132 @@
+            // save recipe
+            JsonNode root = mapper.readTree(json);
+            recipeString = root.path("recipe").toString();                  // select from Json
+            System.out.println(root.path("recipeId").toString());
```
j'image que tu a mis ce system.out pour le debugage. Une fois que ca marche il faut l'enlever

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/recipe/Save.java:
```diff
@@ -0,0 +1,132 @@
+public class Save {
+
+    // java object to json -> take recipe and step from the json -> map to model -> save to db
+    public long saveFunction(Object object, RecipeDao recipeDao, RecipeStepDao recipeStepDao) {
```
je crois que tu essaye de faire de l'unmarshalming ici ?

cocap10 replied:
Je comprend pas a quoi sert cette methode.
Il manque un javadoc qui expliquerai comment tu utilise tes param.
Par exemple object ? j'ai pas l'impression qu'il soit utilisé

cocap10 replied:
RecipeDao recipeDao ? dans Spring tu devrai pouvoir l'autowire

cocap10 replied:
idem pour RecipeStepDao recipeStepDao

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/recipe/Save.java:
```diff
@@ -0,0 +1,132 @@
+                for (JsonNode objectNode : arrayNode) {                     // save each step
+                    i++;
+                    stepString = objectNode.toString();
+                    switch (objectNode.path("selectedStep").textValue()){
```
si c'est un chiffre tu peux peut-etre faire un objectNode.path("selectedStep").intValue() ?

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/recipe/Save.java:
```diff
@@ -0,0 +1,132 @@
+                    stepString = objectNode.toString();
+                    switch (objectNode.path("selectedStep").textValue()){
+                        case "1" :
+                            Sanitizing sanitizing = mapper.readValue(stepString, Sanitizing.class);
```
enfaite tu fais presque la meme chose a chaque etape:
- Tu pourras sortir la ligne "recipeStep.setRecipe(recipe);" de chaque case du switch

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/recipe/Save.java:
```diff
@@ -0,0 +1,132 @@
+
+            // add new steps
+            JsonNode arrayNode = root.path("recipeSteps").path("steps");
+            if (arrayNode.isArray()) {
```
pas de else ? une exception ?

cocap10 commented on src/main/java/org/zenika/zba/zbadata/controller/recipe/Save.java:
```diff
@@ -0,0 +1,132 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Save {
```
c'est un service ?

cocap10 commented on src/main/java/org/zenika/zba/zbadata/dao/RecipeDao.java:
```diff
@@ -1,17 +1,21 @@
 import java.util.List;
 
-@Api(description = "Data Access Object for the user table")
+@SwaggerDefinition(tags = { @Tag(name = "Tag1", description = "Data Access Object for the recipe table") })
```
Tag1 ?

HenriBom committed:
Test : 404 added

cocap10 reviewed (changes requested):
J'ai pas eu le temps de tout relire mais voila mes premiers commentaires

HenriBom committed:
target, out, .mvn removed

cocap10 commented on mvnw:
```diff
@@ -0,0 +1,286 @@
+#!/bin/sh
```
inutile

cocap10 commented on mvnw.cmd:
```diff
@@ -0,0 +1,161 @@
+@REM ----------------------------------------------------------------------------
```
inutile

cocap10 commented on zba-data.iml:
```diff
@@ -0,0 +1,126 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
inutile

cocap10 reviewed (changes requested):
il y a encore des fichier commité pour rien

HenriBom committed:
updated : .gitignore

HenriBom committed:
TS : refacto modele Recipe/Step

- refactoring of the database and model
- recatoring : Set<Step> -> List<Step>

HenriBom committed:
TS : refacto Save class

Refactoring of the Save class to SaveRecipe class as @Service

HenriBom committed:
Test : Controller test updated (in dev)

Still in devellopement

HenriBom committed:
Issue : Deserialization of recipe

Problem in deserializing recipe in post request :
While wanting to send the new recipe and it's step in the same json, step that should point on recipe don't have any id it recipe field yet.

error message : com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `org.zenika.zba.zbadata.model.Recipe` (although at least one Creator exists): no int/Int-argument constructor/factory method to deserialize from Number value (0)

Considered solution : create custom deserialization, seperate deserialization in two parts (recipe then step) allowing to retrive recipeId. Sending the Json in two parts (recipe then step) allowing to retrive recipeId.

HenriBom committed:
Issue : Deserialization of recipe

Problem in deserializing recipe in post request :
While wanting to send the new recipe and it's step in the same json, step that should point on recipe don't have any id it recipe field yet.

error message : com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `org.zenika.zba.zbadata.model.Recipe` (although at least one Creator exists): no int/Int-argument constructor/factory method to deserialize from Number value (0)

Considered solution : create custom deserialization, seperate deserialization in two parts (recipe then step) allowing to retrive recipeId. Sending the Json in two parts (recipe then step) allowing to retrive recipeId.

HenriBom committed:
Issue : Deserialization of recipe

Problem in deserializing recipe in post request :
While wanting to send the new recipe and it's step in the same json, step that should point on recipe don't have any id it recipe field yet.

error message : com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `org.zenika.zba.zbadata.model.Recipe` (although at least one Creator exists): no int/Int-argument constructor/factory method to deserialize from Number value (0)

Considered solution : create custom deserialization, seperate deserialization in two parts (recipe then step) allowing to retrive recipeId. Sending the Json in two parts (recipe then step) allowing to retrive recipeId.

HenriBom committed:
Corrected : Deserialization of recipe

Save class deleted
Deserialization of recipe and steps

HenriBom committed:
Corrected : Deserialization of recipe

Save class deleted
Deserialization of recipe and steps

HenriBom committed:
Merge pull request #3 from Zenika/bug

TS : refacto deserialization of recipe and steps

HenriBom committed:
Refacto : refacto after cocap10  pull request #3 comments

HenriBom committed:
temp commit

HenriBom committed:
Merge pull request #4 from Zenika/temp

temp commit
